File: Lab5/t05_07_e10307.cpp
#include <iostream>
#include <vector>

using namespace std;

bool is_poss_place(int amount, vector<vector<int>>& places, int x) {
    int c = 1;
    int prev_place = places[0][0];
    for (auto seg : places) {
        for (auto i : seg) {
            if (i - prev_place >= x) {
                prev_place = i;
                c += 1;
            }
        }
    }
    return c >= amount;
}

int bin_search(int L, int R, int n, vector<vector<int>>& segs) {
    int l = 0;
    int r = R - L + 1;
    while (r - l > 1) {
        int m = l + (r - l) / 2;
        if (is_poss_place(n, segs, m)) {
            l = m;
        } else {
            r = m;
        }
    }
    return l;
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> segs(m, vector<int>(2));
    for (int i = 0; i < m; i++) {
        cin >> segs[i][0] >> segs[i][1];
    }
    int res = bin_search(segs[0][0], segs[m-1][1], n, segs);
    cout << res << endl;
    return 0;
}

File: Lab5/t05_07_e10307.py
def is_poss_place(amount, places, x):
    c = 1
    prev_place = places[0]
    for i in places:
        if i - prev_place >= x:
            c += 1
            prev_place = i
    return c >= amount


def bin_search(amount, places):
    left, right = 0, places[-1] - places[0] + 1
    while right - left > 1:
        mid = (left + right) // 2
        if is_poss_place(amount, places, mid):
            left = mid
        else:
            right = mid
    return left


if __name__ == "__main__":
    n, m = map(int, input().split())
    places = []
    for i in range(m):
        a, b = map(int, input().split())
        places.append(a)
        places.append(b)
    places.sort()
    res = bin_search(n, places)
    print(res)
